src/EventStore.Core/Services/Archiver/ArchiverService.cs

```diff
@@ -2,54 +2,192 @@
+		_uncommittedChunks = new();
+		_existingChunks = new();
+		_cts = new();
+		_archiveChunkCommands = Channel.CreateUnboundedPrioritized(
+			new UnboundedPrioritizedChannelOptions<Commands.ArchiveChunk> {
+			SingleWriter = true,
+			SingleReader = true,
+			Comparer = new ChunkPrioritizer()
+		});
 
-	public bool Enabled => true;
+		Subscribe();
+		Start();
+	}
+
+	private void Subscribe() {
+		_mainBus.Subscribe<SystemMessage.ChunkLoaded>(this);
+		_mainBus.Subscribe<SystemMessage.ChunkSwitched>(this);
+		_mainBus.Subscribe<SystemMessage.ChunkCompleted>(this);
+		_mainBus.Subscribe<ReplicationTrackingMessage.ReplicatedTo>(this);
+		_mainBus.Subscribe<SystemMessage.SystemStart>(this);
+		_mainBus.Subscribe<SystemMessage.BecomeShuttingDown>(this);
+	}
+
+	private void Start() {
+		Task.Factory.StartNew(
+			() => ArchiveChunks(_cts.Token),
+			_cts.Token,
+			TaskCreationOptions.LongRunning,
+			TaskScheduler.Current);
+	}
+
+	public void Handle(SystemMessage.ChunkLoaded message) {
+		if (!message.ChunkInfo.IsCompleted)
+			return;
 
-	public string LicensePublicKey => LicenseConstants.LicensePublicKey;
+		_existingChunks[Path.GetFileName(message.ChunkInfo.ChunkFileName)!] = message.ChunkInfo;
+	}
 
-	public void ConfigureApplication(IApplicationBuilder builder, IConfiguration configuration) {
-		var licenseService = builder.ApplicationServices.GetRequiredService<ILicenseService>();
-		_ = LicenseMonitor.MonitorAsync(
-			featureName: Name,
-			requiredEntitlements: ["ARCHIVE"],
-			licenseService: licenseService,
-			onLicenseException: licenseService.RejectLicense,
-			logger: builder.ApplicationServices.GetRequiredService<ILoggerFactory>().CreateLogger(GetType()));
+	public void Handle(SystemMessage.ChunkCompleted message) {
+		var chunkInfo = message.ChunkInfo;
+		if (chunkInfo.ChunkEndPosition > _replicationPosition) {
+			_uncommittedChunks.Enqueue(chunkInfo);
+			return;
+		}
 
-		var options = configuration.GetSection("EventStore:Archive").Get<ArchiverOptions>();
-		var storage = CreateStorage(options);
+		ScheduleChunkForArchiving(chunkInfo, "new");
 	}
 
-	public void ConfigureServices(IServiceCollection services, IConfiguration configuration) {
+	public void Handle(SystemMessage.ChunkSwitched message) {
+		ScheduleChunkForArchiving(message.ChunkInfo, "changed");
 	}
 
-	static IArchiveStorage CreateStorage(ArchiverOptions options) =>
-		options.StorageType switch {
-			StorageType.None => IArchiveStorage.None,
-			StorageType.FileSystem => new FileSystemArchiveStorage(options.FileSystem),
-			StorageType.S3 => new S3ArchiveStorage(options.S3),
-			_ => throw new InvalidOperationException(),
-		};
+	public void Handle(ReplicationTrackingMessage.ReplicatedTo message) {
+		_replicationPosition = Math.Max(_replicationPosition, message.LogPosition);
+		ProcessUncommittedChunks();
+	}
+
+	public void Handle(SystemMessage.SystemStart message) {
+		Task.Factory.StartNew(() => ScheduleExistingChunksForArchiving(_cts.Token), _cts.Token);
```

Comment:
task.run will do here for simplicity (https://blog.stephencleary.com/2013/08/startnew-is-dangerous.html)

Reply:
fixed